# Problem Set 4 Yunqiu(Julie) Li

In [1]:
import pandas as pd
import numpy as np
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.svm import LinearSVC
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier

In [2]:
# load default data set
credit = pd.read_csv("defaultBal.csv")
# set up data for scikit-learn
# note:  You don't really have to use X,y
# this is our target
y = credit.default
# These are predictors (skip id)
Xall= credit.values[:,1:24]
# copy of all data to predictor
X = Xall.copy()
# Set up as just real values, use this for last part of problem
# Restricts predictor to only real values (not discrete)
# X = Xall[:,12:23].copy()
print(X.shape)
print(np.mean(y))
yGuess = np.mean(y)

(12450, 23)
0.5330923694779116


In [3]:
credit.head()

id  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0   1      20000    2          2         1   24      2      2     -1     -1   
1   1      20000    2          2         1   24      2      2     -1     -1   
2   2     120000    2          2         2   26     -1      2      0      0   
3   6      50000    1          1         2   37      0      0      0      0   
4   9     140000    2          3         1   28      0      0      2      0   

    ...     BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0   ...             0          0          0         0       689         0   
1   ...             0          0          0         0       689         0   
2   ...          3272       3455       3261         0      1000      1000   
3   ...         19394      19619      20024      2500      1815       657   
4   ...         12211      11793       3719      3329         0       432   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  default  
0         0         0         0        1  
1         0         0         0        1  
2      1000         0      2000        1  
3      1000      1000       800        0  
4      1000      1000      1000        0  

[5 rows x 25 columns]

In [4]:
# Note:Use test_size=0.25 throughout
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)

##  Naive Bayes example

In [5]:
# Try a naive Bayes classifier
gnb = GaussianNB()
trainFit = gnb.fit(X_train,y_train)
print(trainFit.score(X_train,y_train))
print(trainFit.score(X_test, y_test))

0.566241833565385
0.5778991326694507


In [8]:
# start monte-carlo for GaussianNB()
nmc = 50
trainScore = np.zeros(nmc)
testScore  = np.zeros(nmc)
gnb = GaussianNB()
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    trainFit = gnb.fit(X_train,y_train)
    trainScore[i] = trainFit.score(X_train,y_train)
    testScore[i] =  trainFit.score(X_test,y_test)
print(np.mean(trainScore))
print(np.std(trainScore))
print(np.mean(testScore))
print(np.std(trainScore))
print(np.mean(testScore>yGuess))

0.5658391346256828
0.0031036734810529288
0.5644201734661098
0.0031036734810529288
1.0


### Problem 1: Repeat the monte-carlo with a LinearDiscriminant classifier. Report means, std’s, and probability of beating a benchmark classifier which guesses 1 all the time.

In [23]:
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    lda = LinearDiscriminantAnalysis().fit(X_train,y_train)
    trainScore[i] = lda.score(X_train,y_train)
    testScore[i] =  lda.score(X_test,y_test)
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

Training accuracy : 0.575
Training standard deviation : 0.005
Testing accuracy : 0.572
Testing standard deviation : 0.008
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000


### Problem 2: Repeat this with Logistic regression for C=100, C=1, C=0.01.

In [24]:
### run regression individually for difference values of C instead of using for loop, because running for loop would long time.
C = 100
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    LR = LogisticRegression(C=C).fit(X_train,y_train)
    trainScore[i] = LR.score(X_train,y_train)
    testScore[i] =  LR.score(X_test,y_test)
print("When C = " + str(C))
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

When C = 100
Training accuracy : 0.585
Training standard deviation : 0.006
Testing accuracy : 0.583
Testing standard deviation : 0.009
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000


In [25]:
C = 1
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    LR = LogisticRegression(C=C).fit(X_train,y_train)
    trainScore[i] = LR.score(X_train,y_train)
    testScore[i] =  LR.score(X_test,y_test)
print("When C = " + str(C))
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".
      format(np.mean(testScore>yGuess)))

When C = 1
Training accuracy : 0.584
Training standard deviation : 0.007
Testing accuracy : 0.583
Testing standard deviation : 0.008
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000


In [26]:
C = 0.01
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    LR = LogisticRegression(C=0.01).fit(X_train,y_train)
    trainScore[i] = LR.score(X_train,y_train)
    testScore[i] =  LR.score(X_test,y_test)
print("When C = " + str(C))
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

When C = 0.01
Training accuracy : 0.584
Training standard deviation : 0.008
Testing accuracy : 0.583
Testing standard deviation : 0.008
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000


### Problem 3: Repeat this for a LinearSVC with C = 100, C=1, C=0.01

In [27]:
C = 100
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    LinSVC = LinearSVC(C=C).fit(X_train,y_train)
    trainScore[i] = LinSVC.score(X_train,y_train)
    testScore[i] =  LinSVC.score(X_test,y_test)
print("When C = " + str(C))
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

When C = 100
Training accuracy : 0.531
Training standard deviation : 0.036
Testing accuracy : 0.530
Testing standard deviation : 0.037
Probability of beating a benchmark classifier which guesses 1 all the time : 0.580


In [28]:
C = 1
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    LinSVC = LinearSVC(C=C).fit(X_train,y_train)
    trainScore[i] = LinSVC.score(X_train,y_train)
    testScore[i] =  LinSVC.score(X_test,y_test)
print("When C = " + str(C))
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

When C = 1
Training accuracy : 0.526
Training standard deviation : 0.035
Testing accuracy : 0.528
Testing standard deviation : 0.038
Probability of beating a benchmark classifier which guesses 1 all the time : 0.500


In [29]:
C = 0.01
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    LinSVC = LinearSVC(C=C).fit(X_train,y_train)
    trainScore[i] = LinSVC.score(X_train,y_train)
    testScore[i] =  LinSVC.score(X_test,y_test)
print("When C = " + str(C))
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

When C = 0.01
Training accuracy : 0.528
Training standard deviation : 0.043
Testing accuracy : 0.526
Testing standard deviation : 0.042
Probability of beating a benchmark classifier which guesses 1 all the time : 0.560


### Problem 4: Repeat this for KNeighborClassifier for n_neighbors = 3, 11, 25


In [16]:
n_neighbors = 3
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    KNN = KNeighborsClassifier(n_neighbors=n_neighbors).fit(X_train,y_train)
    trainScore[i] = KNN.score(X_train,y_train)
    testScore[i] =  KNN.score(X_test,y_test)
print("When n_neighbors = " + str(n_neighbors))
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

When n_neighbors = 3
Training accuracy : 0.779
Training standard deviation : 0.003
Testing accuracy : 0.570
Testing standard deviation : 0.009
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000


In [17]:
n_neighbors = 11
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    KNN = KNeighborsClassifier(n_neighbors=n_neighbors).fit(X_train,y_train)
    trainScore[i] = KNN.score(X_train,y_train)
    testScore[i] =  KNN.score(X_test,y_test)
print("When n_neighbors = " + str(n_neighbors))
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

When n_neighbors = 11
Training accuracy : 0.680
Training standard deviation : 0.003
Testing accuracy : 0.601
Testing standard deviation : 0.008
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000


In [18]:
n_neighbors = 25
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    KNN = KNeighborsClassifier(n_neighbors=n_neighbors).fit(X_train,y_train)
    trainScore[i] = KNN.score(X_train,y_train)
    testScore[i] =  KNN.score(X_test,y_test)
print("When n_neighbors = " + str(n_neighbors))
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

When n_neighbors = 25
Training accuracy : 0.650
Training standard deviation : 0.003
Testing accuracy : 0.608
Testing standard deviation : 0.008
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000


### Problem 5: Repeat this for a Decision Tree with max_depth = 5, 10, 25


In [19]:
for max_depth in [5,10,25]:
    for i in range(nmc):
        X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
        tree = DecisionTreeClassifier(max_depth=4).fit(X_train,y_train)
        trainScore[i] = tree.score(X_train,y_train)
        testScore[i] =  tree.score(X_test,y_test)
    print("When max_depth = " + str(max_depth))
    print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
    print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
    print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
    print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
    print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

When max_depth = 5
Training accuracy : 0.697
Training standard deviation : 0.003
Testing accuracy : 0.688
Testing standard deviation : 0.009
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000
When max_depth = 10
Training accuracy : 0.698
Training standard deviation : 0.004
Testing accuracy : 0.689
Testing standard deviation : 0.008
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000
When max_depth = 25
Training accuracy : 0.698
Training standard deviation : 0.004
Testing accuracy : 0.686
Testing standard deviation : 0.009
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000


### Prblem 6: Repeat the linear LinearDiscriminant, but use just the real valued data. See example file for line to move just these fields into X.


In [20]:
# Restricts predictor to only real values (not discrete)
X = Xall[:,12:23].copy()
for i in range(nmc):
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)
    lda = LinearDiscriminantAnalysis().fit(X_train,y_train)
    trainScore[i] = lda.score(X_train,y_train)
    testScore[i] =  lda.score(X_test,y_test)
print("Training accuracy : {:.3f}".format(np.mean(trainScore)))
print("Training standard deviation : {:.3f}".format(np.std(trainScore)))
print("Testing accuracy : {:.3f}".format(np.mean(testScore)))
print("Testing standard deviation : {:.3f}".format(np.std(testScore)))
print("Probability of beating a benchmark classifier which guesses 1 all the time : {:.3f}".format(np.mean(testScore>yGuess)))

Training accuracy : 0.575
Training standard deviation : 0.004
Testing accuracy : 0.573
Testing standard deviation : 0.009
Probability of beating a benchmark classifier which guesses 1 all the time : 1.000


### Problem 7: Using the mean test score from parts (1-5+GaussianNB), what was your best method (so far)?


Decision Tree was my best method, because it has the best fit (highest score - approximate 0.68) for test data among all the classification methods, and there won't be overfitting problem (small gap bettwen train score and train score - train score : 0.69 vs. test score : 0.68).  